core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java

```diff
@@ -561,6 +593,234 @@ public UpdateNode plan(Update node)
+
+            Optional<Expression> rewritten = mergeCase.getExpression().map(subPlan::rewrite);
+            Expression condition = presentColumn.toSymbolReference();
+            if (mergeCase instanceof MergeInsert) {
+                condition = new IsNullPredicate(presentColumn.toSymbolReference());
+            }
+
+            if (rewritten.isPresent()) {
+                condition = ExpressionUtils.and(condition, rewritten.get());
+            }
+
+            whenClauses.add(new WhenClause(condition, new Row(rowBuilder.build())));
+        }
+
+        ImmutableList.Builder<Expression> rowBuilder = ImmutableList.builder();
+        mergeAnalysis.getDataColumnSchemas().forEach(columnSchema ->
+                rowBuilder.add(new Cast(new NullLiteral(), toSqlType(columnSchema.getType()))));
+        rowBuilder.add(new IsNotNullPredicate(presentColumn.toSymbolReference()));
+        rowBuilder.add(new GenericLiteral("INTEGER", "-1"));
+
+        SearchedCaseExpression caseExpression = new SearchedCaseExpression(whenClauses.build(), Optional.of(new Row(rowBuilder.build())));
+        RowType rowType = createRowType(mergeAnalysis.getDataColumnSchemas());
+        Expression castCaseExpression = new Cast(caseExpression, toSqlType(rowType));
+
+        Assignments.Builder projectionAssignmentsBuilder = Assignments.builder();
+        for (ColumnHandle column : mergeAnalysis.getRedistributionColumnHandles()) {
+            int fieldIndex = requireNonNull(mergeAnalysis.getColumnHandleFieldNumbers().get(column), "Could not find fieldIndex for redistribution column");
+            Field field = target.getDescriptor().getFieldByIndex(fieldIndex);
+            Symbol symbol = symbolAllocator.newSymbol(field.getName().get(), field.getType());
+            projectionAssignmentsBuilder.put(symbol, target.getFieldMappings().get(fieldIndex).toSymbolReference());
+        }
+
+        FieldReference reference = analysis.getRowIdField(mergeAnalysis.getTargetTable());
+        Field rowIdField = target.getDescriptor().getFieldByIndex(reference.getFieldIndex());
+        Symbol rowId = symbolAllocator.newSymbol("row_id", rowIdField.getType());
+        projectionAssignmentsBuilder.put(rowId, target.getFieldMappings().get(reference.getFieldIndex()).toSymbolReference());
+        Symbol mergeOutput = symbolAllocator.newSymbol("merge_row", rowType);
+        projectionAssignmentsBuilder.put(mergeOutput, castCaseExpression);
+
+        ProjectNode project = new ProjectNode(
+                idAllocator.getNextId(),
+                subPlan.getRoot(),
+                projectionAssignmentsBuilder.build());
+
+        // Projecting the writeRedistributionColumns, the rowId block, and the merge RowBlock
+        int expectedSize = redistributionColumns.size() + 1 + 1;
+        int actualSize = project.getOutputSymbols().size();
+        checkArgument(actualSize == expectedSize, "projectedSymbols should have size %s, but is %s", expectedSize, actualSize);
```

Comment:
And there can't be duplicates in the `redistributionColumns` because they are a subset of the table columns.  Good point, this check is unnecessary, so I deleted it.